### PR TITLE
Changes in Dynamic Modules

### PR DESCRIPTION
**Pull request by AlbinSou**
I tried to implement the changes that would fix bugs described in Issue #1591 and also allow users to use Dynamic modules more easily outside of avalanche strategies.

To do so, I made several major changes.

- Remove old avalanche_model_adaptation method and replace it with a new version which does things differently. Instead of iterating on all modules, it iterates only on childrens, and does so recursively in order to explore the whole hierarchy. It still iterates on the hierarchy of nn.Modules and catches DynamicModules to adapt them just like before.
- Added a method adapt() to dynamicmodule that will call the new "avalanche_model_adaptation" on itself and recursively on it's submodules
- Added an _auto_adapt attribute to all DynamicModules that triggers or not the adaptation inside the avalanche_model_adaptation recursion (defaults to True). In particular, this attribute does not prevent the user to call directly module.adapt() if he wants, however if the module needs to be adapted as a second level or more module inside this recursion, it will not get adapted. This is the part that fixes the bug described in #1591, since that way we can instantiate IncrementalClassifiers inside the MultiHeadClassifier that will not be automatically adapted but rather adapted manually by the MultiHeadClassifier only if the experience used for adaptation contains the task id related to this IncrementalClassifier.
- Made a few clarification changes in IncrementalClassifier and Multiheadclassifier by replacing occurences of adaptation() implementation by train_adaptation and eval_adaptation, where if the eval_adaptation needs to be the same than train_adaptation I just call train_adaptation inside eval_adaptation (nested adaptation() override were a bit complex to understand and could lead to confusion)

For now tests do not pass, but some of them I am not sure what behaviour they were checking and if we really want this behaviour. I will also add some more tests to check that the behaviour in #1591 is solved (my local test says it is fixed for now).

### CODE REVIEW

File: tests/models/test_models.py [hunk 1]
```diff
@@ -455,13 +456,23 @@ def test_multihead_head_selection(self):
         for x, y, t in DataLoader(benchmark.train_stream[0].dataset):
             y_mh = model(x, t)
             y_t = model_t0(x)
+
+            # We need to pad y_t to dim with zeros
+            # because y_mh will have max dim of all heads
+            y_t = F.pad(y_t, (0, y_mh.size(1) - y_t.size(1)))
+
```
Comment (AlbinSou) [inline tests/models/test_models.py:463]: Previous test was assuming that y_t and y_mh are always of the same size. This is not true. The reason why is that when passing a vector as a target to MultiHeadClassifier, it will always return an output that is padded to the maximum size of one of it's head, this is because it assumes that more than one unique target can be inside the vector (unfortunately, most of the time we indeed only pass one task inside the vector, maybe this is a behaviour we want to change). However, because of the bug the output y_t was also of the size of the biggest task because all IncrementalClassifiers were updated on all experiments. Now it's not the case anymore so before checking equality I have to pad y_t with zeros.

File: avalanche/models/dynamic_modules.py [hunk 2]
```diff
@@ -12,12 +12,41 @@
...
+    _visited=None,
+    _initial_call: bool = True,
+):
+    if _visited is None:
+        _visited = []
+
+    if module in _visited:
+        return
+
+    _visited.append(module)
```
Comment (AntonioCarta) [inline avalanche/models/dynamic_modules.py:36]: it is better if `_visited` is a set for efficiency in the search.

File: avalanche/models/dynamic_modules.py [hunk 3]
```diff
@@ -12,12 +12,41 @@
...
-from avalanche.benchmarks.utils.flat_data import ConstantSequence
 from avalanche.benchmarks.scenarios import CLExperience
+from avalanche.benchmarks.utils.flat_data import ConstantSequence
+
+
+def avalanche_model_adaptation(
+    module: Module,
+    experience: CLExperience,
+    _visited=None,
+    _initial_call: bool = True,
```
Comment (AntonioCarta) [inline avalanche/models/dynamic_modules.py:28]: can you add a comment about why we need `_initial_call`?

File: tests/models/test_models.py [hunk 4]
```diff
@@ -416,8 +417,8 @@ def test_multihead_head_creation(self):
             w.data_ptr() for group in optimizer.param_groups for w in group["params"]
         ]
 
-        assert w_ptr not in opt_params_ptrs  # head0 has been updated
-        assert b_ptr not in opt_params_ptrs  # head0 has been updated
+        # assert w_ptr not in opt_params_ptrs  # head0 has NOT been updated
+        # assert b_ptr not in opt_params_ptrs  # head0 has NOT been updated
```
[the remark this answers is not in the document]
Comment (AntonioCarta) [reply tests/models/test_models.py:421]: Are you sure? I think it's if head0 is changed by SGD when training on task 4, which should not happen.

File: tests/models/test_models.py (hunk 1 above)
Comment (AntonioCarta) [reply tests/models/test_models.py:463]: Correct.

**Comment by AntonioCarta** [comment]
> Made a few clarification changes in IncrementalClassifier and Multiheadclassifier by replacing occurences of adaptation() implementation by train_adaptation and eval_adaptation, where if the eval_adaptation needs to be the same than train_adaptation I just call train_adaptation inside eval_adaptation (nested adaptation() override were a bit complex to understand and could lead to confusion)

I think we could also remove `train_adaptation` and `eval_adaptation` and have a single method. After, you can detect train/eval mode with a single if when necessary, so it seems to be adding unnecessary complexity in the common case.

File: tests/models/test_models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlbinSou) [reply tests/models/test_models.py:421]: w_ptr contains the head0 data pointer after head0 has been adapted on task 0. This pointer should still be the one present in optimizer after adaptation is done on task 4, since it should not be touched by the task 4 adaptation (which modifies head 4 only, and not head 0).

**Comment by AlbinSou** [comment]
@AntonioCarta  How would you like to manage the adapt() and adaptation() function ? I wanted to keep the name of the "adaptation" also to keep backward compatibility but now it's a bit confusing to have two methods adapt() and adaptation()

**Comment by AntonioCarta** [comment]
> @AntonioCarta How would you like to manage the adapt() and adaptation() function ? I wanted to keep the name of the "adaptation" also to keep backward compatibility but now it's a bit confusing to have two methods adapt() and adaptation()

The advantage of the current solution is that it's backward compatible. If we don't want to rename the old method we could rename the new one as `rec_adapt` for recursive adaptation?

I was also thinking about a test case to check that the recursive calls don't break the #1591 fix:
```
model = ABasicDynamicModule(ABasicModule(MultiTaskClassifier(...)))
model.adapt(exp)  # here it's calling rec_adapt from the "base" layer, not the MultiTaskClassifier. It should still work.
# assert that wrong heads are not adapted
```

**Comment by AlbinSou** [comment]
> > @AntonioCarta How would you like to manage the adapt() and adaptation() function ? I wanted to keep the name of the "adaptation" also to keep backward compatibility but now it's a bit confusing to have two methods adapt() and adaptation()
> 
> The advantage of the current solution is that it's backward compatible. If we don't want to rename the old method we could rename the new one as `rec_adapt` for recursive adaptation?
> 
> I was also thinking about a test case to check that the recursive calls don't break the #1591 fix:
> 
> ```
> model = ABasicDynamicModule(ABasicModule(MultiTaskClassifier(...)))
> model.adapt(exp)  # here it's calling rec_adapt from the "base" layer, not the MultiTaskClassifier. It should still work.
> # assert that wrong heads are not adapted
> ```

I added a test for that at the end of test_models.DynamicModulesTest. Yes, maybe I will rename to make it clearer

File: tests/models/test_models.py [hunk 5]
```diff
@@ -511,6 +522,42 @@ def test_incremental_classifier_update_masking_only_during_training(self):
...
+        for t, exp in enumerate(benchmark.train_stream):
+            avalanche_model_adaptation(model, exp)
+
+        for t, s in sizes.items():
+            self.assertEqual(s, model.classifiers[str(t)].classifier.out_features)
+
+    def test_avalanche_adaptation(self):
+        model1 = torch.nn.Sequential(MultiHeadClassifier(in_features=6))
+        benchmark = get_fast_benchmark(use_task_labels=True, shuffle=True)
+        avalanche_model_adaptation(model1, benchmark.train_stream[0])
```
Comment (AntonioCarta) [inline tests/models/test_models.py:543]: the test should check that only the correct modules are updated in this test case

Comment (AlbinSou) [reply tests/models/test_models.py]: I was doing it in a test above, I changed them a bit now I have two tests, one for avalanche_model_adaptation() and one for recursive_adaptation, both test that the correct modules were adapted (through checking that the sizes of the heads do not change when doing 2 adaptations). I removed the first test_multihead_size because it was doing the same.

**Comment by AlbinSou** [comment]
@AntonioCarta On my side everything is ok. I was thinking maybe to include an example on how to use dynamic modules outside and inside of a strategy (call to recursive_adaptation and explanation on how avalanche_model_adaptation works). Another thing is that I changed the location of avalanche_model_adaptation for now it's inside dynamic_modules.py before it was in models/utils.py. For now I import it in utils so that ppl don't have to change their imports, but it's a bit weird. The reason I put it inside dynamic_modules is that this function is needed in recursive_adaptation() method so putting it to utils was leading to annoying recursive imports.

**Comment by AntonioCarta** [comment]
Ok. I will wait to merge this after release to be on the safe side.

> I was thinking maybe to include an example on how to use dynamic modules outside and inside of a strategy 

we already have examples of that in the notebooks. Maybe you can just update them?
